8253742: POSIX signal code cleanup

gerard-ziemski opened the pull request:
hi all,

Please review this followup to [JDK-8252324 Signal related code should be shared among POSIX platforms](https://bugs.openjdk.java.net/browse/JDK-8252324), where several issues were identified for a cleanup. This change addresses them all:

#1 David's feedback - removed non POSIX SIGNIFICANT_SIGNAL_MASK code

#2 David's feedback - used unblock_program_error_signals() on all platforms (reverted for JDK-8252533)

#3 David's feedback - used single JVM_handle_posix_signal API for all POSIX platforms (reverted for JDK-8255711)

#4 Coleen's feedback - cleanup header files in src/hotspot/os/posix/signals_posix.hpp

#5 Coleen's feedback - hid SR_signum assignment in src/hotspot/os/posix/signals_posix.hpp to avoid having to include <signal.h>

#6 Coleen's feedback - factored out print_signal_handlers()

#7 Thomas' feedback - factored out common POSIX os::SuspendedThreadTask::internal_do_task()

#8 Thomas's feedback - factored out common POSIX signal initialization code

#9 YaSuenag's feedback - used JVM_handle_posix_signal for the common API

#10 YaSuenag's feedback - unified logging out of the scope for this fix

#11 YaSuenag's feedback - memset usage in PosixSignals::jdk_misc_signal_init() correct?

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8253742](https://bugs.openjdk.java.net/browse/JDK-8253742): POSIX signal code cleanup


### Reviewers
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**)
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/636/head:pull/636`
`$ git checkout pull/636`

magicus commented:
/label remove build

gerard-ziemski commented:
Does anyone have a suggestion on how to merge with current JDK in the safest way?

tstuefe commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -436,70 +432,34 @@ bool PosixSignals::chained_handler(int sig, siginfo_t* siginfo, void* context) {
-// Set thread signal mask (for some reason on AIX sigthreadmask() seems
-// to be the thing to call; documentation is not terribly clear about whether
-// pthread_sigmask also works, and if it does, whether it does the same.
-bool set_thread_signal_mask(int how, const sigset_t* set, sigset_t* oset) {
```
As remarked above, this will conflict with JDK-8252533, since I remove this function too. Can we leave this out please?

gerard-ziemski replied:
Reverted.

tstuefe commented on src/hotspot/share/runtime/os.hpp:
```diff
@@ -966,6 +966,8 @@ class os: AllStatic {
 #endif // !WINDOWS
 
+  static address ucontext_get_pc(const ucontext_t* ctx);
+  static void    ucontext_set_pc(ucontext_t* ctx, address pc);
```
This feels misplaced here (and probably won't compile on windows) since ucontext_t is POSIX. At the very least needs ucontext.h. But I would consider moving this to os_posix.

gerard-ziemski replied:
I thought I tested it and it built fine on Windows - will take another look...

gerard-ziemski replied:
Fixed.

tstuefe reviewed (commented):
Hi Gerard,

good job!

But we really should synchronize  :)

I am currently working on: https://bugs.openjdk.java.net/browse/JDK-8255711
(see Draft PR: https://github.com/openjdk/jdk/pull/982)

and https://bugs.openjdk.java.net/browse/JDK-8252533 is also still open (see https://github.com/openjdk/jdk/pull/839) - still waiting Davids final OK.

So unfortunately there are a number of clashes with your change:

- "JVM_handle_xxx_signal()": See my mail to jdk-dev from this morning: https://mail.openjdk.java.net/pipermail/jdk-dev/2020-November/004887.html - we either can get completely rid of this function, which is what my current Draft for JDK-8255711 does. Or we need to retain it for backward compatibility, but if so, we need to retain it with the current interface.

Either way, could you please withhold changes to the hotspot signal handlers for the moment (so, both javaSignalHandler() and the various JVM_handle_xxx_signal() functions)?

- I removed some functions you changed - all that signal blocking mask stuff. See https://github.com/openjdk/jdk/pull/839. Could you hold any changes to those functions until JDK-8252533 is out of the door? Hope to do this tomorrow, with your and Davids approval.

The unification of the signal handler printing stuff and the SR initialization make sense and are a good simplification.
 
Please fine further remarks remarks inline.

Cheers, Thomas

dholmes-ora commented:
I hadn't realized that JVM_handle_XXX_signal defined a per-platform "public" entry point to allow external callers of the signal handling function in conjunction with -XX:+AllowUserSignalHandlers. We need to keep these but they can each call JVM_handle_posix_signal as their implementation.

tstuefe commented:
> I hadn't realized that JVM_handle_XXX_signal defined a per-platform "public" entry point to allow external callers of the signal handling function in conjunction with -XX:+AllowUserSignalHandlers. We need to keep these but they can each call JVM_handle_posix_signal as their implementation.

We should disentangle https://bugs.openjdk.java.net/browse/JDK-8255711 and this patch, https://bugs.openjdk.java.net/browse/JDK-8253742.

I started by giving my patch a less generic name ("Fix and unify hotspot signal handlers"). I propose to do the same with this patch, or even split this patch into two smaller parts, since it does two things:
- unify diagnostic printing code
- unify SR handler setup

As I wrote, I'd prefer to keep changes to JVM_xxx and javaSignalHandler out of this patch completely. I have to change those functions since the point of my patch is signal handler unification. 

In turn, I will keep my hands off any other code in signals_posix.xxx to decrease chances of conflict with this patch.

tstuefe commented:
> > I hadn't realized that JVM_handle_XXX_signal defined a per-platform "public" entry point to allow external callers of the signal handling function in conjunction with -XX:+AllowUserSignalHandlers. We need to keep these but they can each call JVM_handle_posix_signal as their implementation.
> 
> We should disentangle https://bugs.openjdk.java.net/browse/JDK-8255711 and this patch, https://bugs.openjdk.java.net/browse/JDK-8253742.
> 
> I started by giving my patch a less generic name ("Fix and unify hotspot signal handlers"). I propose to do the same with this patch, or even split this patch into two smaller parts, since it does two things:
> 
> * unify diagnostic printing code
> * unify SR handler setup
> 
> As I wrote, I'd prefer to keep changes to JVM_xxx and javaSignalHandler out of this patch completely. I have to change those functions since the point of my patch is signal handler unification.
> 
> In turn, I will keep my hands off any other code in signals_posix.xxx to decrease chances of conflict with this patch.

Oh, and yes, I preserve the JVM_handle_xxx_signal entries in my patch, but they are thin wrappers around an internal, posix-specific handler function.

gerard-ziemski commented on make/hotspot/symbols/symbols-linux:
```diff
@@ -21,7 +21,7 @@
 #
 
-JVM_handle_linux_signal
+JVM_handle_posix_signal
```
Reverted.

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -436,70 +432,34 @@ bool PosixSignals::chained_handler(int sig, siginfo_t* siginfo, void* context) {
-extern "C" JNIEXPORT int JVM_handle_linux_signal(int signo, siginfo_t* siginfo,
-                                               void* ucontext,
-                                               int abort_if_unrecognized);
-#endif
```
Reverted.

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1291,6 +1245,28 @@ void PosixSignals::print_signal_handler(outputStream* st, int sig,
+  PosixSignals::print_signal_handler(st, SHUTDOWN2_SIGNAL , buf, buflen);
+  PosixSignals::print_signal_handler(st, SHUTDOWN3_SIGNAL , buf, buflen);
+  PosixSignals::print_signal_handler(st, BREAK_SIGNAL, buf, buflen);
+#if defined(AIX)
```
Fixed.

gerard-ziemski commented on src/hotspot/os/aix/os_aix.cpp:
```diff
@@ -1440,22 +1440,7 @@ void os::pd_print_cpu_info(outputStream* st, char* buf, size_t buflen) {
 }
 
 void os::print_signal_handlers(outputStream* st, char* buf, size_t buflen) {
```
print_signal_handlers() is declared in os.h, so at first glance I thought I couldn't collapse it further. I'll have a second look.

gerard-ziemski replied:
void os::print_signal_handlers(outputStream* st, char* buf, size_t buflen) is declared in os.h and is used on Windows platform as well.

dholmes-ora replied:
So we have one definition in os_posix.cpp and one in os_windows.cpp.

gerard-ziemski commented:
The snapshot of JDK that I'm using in my PR does not build on Windows. Do you have any suggestion how I can safely update to the latest JDK without messing up my PR?

dholmes-ora commented on src/hotspot/os/aix/os_aix.cpp:
```diff
@@ -2600,11 +2576,7 @@ bool os::bind_to_processor(uint processor_id) {
 }
 
 void os::SuspendedThreadTask::internal_do_task() {
```
We should be able to have a single definition of this function in os_posix.cpp too.

gerard-ziemski replied:
Fixed.

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1291,6 +1281,29 @@ void PosixSignals::print_signal_handler(outputStream* st, int sig,
 
+void PosixSignals::print_signal_handlers(outputStream* st, char* buf, size_t buflen) {
+  st->print_cr("Signal Handlers:");
+  PosixSignals::print_signal_handler(st, SIGSEGV, buf, buflen);
```
You shouldn't need the PosixSignals:: prefix in this method.

gerard-ziemski replied:
Goo catch, fixed.

gerard-ziemski replied:
Actually now that we are changing `PosixSignals::print_signal_handlers()` to `os::print_signal_handlers()` we do need the `PosixSignals::` prefix after all.

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1357,7 +1346,7 @@ void PosixSignals::signal_sets_init() {
     sigaddset(&unblocked_sigs, SIGTRAP);
   #endif
-  sigaddset(&unblocked_sigs, SR_signum);
+  sigaddset(&unblocked_sigs, PosixSignals::SR_signum);
```
Shouldn't need the PosixSignals::prefix in this method

gerard-ziemski replied:
Here we are going from:
`void PosixSignals::signal_sets_init()`
to
` void signal_sets_init()`
so in this case we do need it.

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1648,3 +1637,30 @@ void PosixSignals::do_resume(OSThread* osthread) {
 }
+
+void PosixSignals::do_task(Thread* thread, os::SuspendedThreadTask* task) {
+  if (PosixSignals::do_suspend(thread->osthread())) {
```
Shouldn't need PosixSignals:: prefix in this method.

gerard-ziemski replied:
Leftover from the time when do_task wasn't in signals_posix.cpp, fixed.

dholmes-ora commented on src/hotspot/os/posix/signals_posix.hpp:
```diff
@@ -25,45 +25,45 @@
+// Forward declarations to be independent of the include structure.
 
-// Signal number used to suspend/resume a thread
-// do not use any signal number less than SIGSEGV, see 4355769
-static int SR_signum = SIGUSR2;
+typedef siginfo_t siginfo_t;
+typedef sigset_t sigset_t;
```
I don't see why this is needed/wanted. We can include signal.h without a problem.

I'm not even sure what these typedefs means ??

gerard-ziemski replied:
Coleen asked me to remove <signal.h> from the signals_posix.hpp, so those are forward declarations for the signal types we use.

I thought it was a reasonable request to minimize the number of headers. I saw some efforts in the past to cleanup header files, which is supposed to help with build times, so every little bit helps.

tstuefe replied:
Hi Gerard,

the proper way to do this would be to remove the header and forward declare the structures like this:

```
struct sigset_t;
struct siginfo_t;
struct ucontext_t;
```
That means we can use pointers to those things without including their definition.

But in this case I believe you could just remove signal.h without forward declaring anything. System headers are usually included via globalDefinitions.hpp, which we do include here. See globalDefinitions_gcc.hpp in this case.

Cheers, Thomas

tstuefe replied:
But its actually a matter of taste; leaving the forward declarations explicitly in like described above would be okay too.

dholmes-ora reviewed (commented):
Hi Gerard,

Overall looking good. Some changes still to be finalized e.g ucontext_t related functions in os.hpp.

I flagged some os functions that are implemented on os_foo.cpp but which just call the Posix helper, which can be deleted from os_foo.cpp and simply added to os_posix.cpp. That can't be a further cleanup RFE if you want to limit changes in this PR.

A few minor nits below.

Thanks,
David

tstuefe commented:
> The snapshot of JDK that I'm using in my PR does not build on Windows. Do you have any suggestion how I can safely update to the latest JDK without messing up my PR?

Hi Gerard,

merging master would not help you with the build error. As I said, it complains about ucontext_t. 

As for your question, just do a 

```
git checkout yourbranch
git merge master
```

if you get conflicts, you'll need to resolve them, but this is the way to go without invalidating old commits.

tstuefe commented:
> _Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [hotspot-dev](mailto:hotspot-dev@openjdk.java.net):_
> 
> On 7/11/2020 2:40 am, Gerard Ziemski wrote:
> 
> > On Wed, 4 Nov 2020 04:22:05 GMT, David Holmes <dholmes at openjdk.org> wrote:
> > > > Gerard Ziemski has updated the pull request incrementally with two additional commits since the last revision:
> > > > - use ifdef(SIGDANGER) and ifdef(SIGTRAP)
> > > > - revert unblock_program_error_signals change
> > > 
> > > 
> > > src/hotspot/os/posix/signals_posix.hpp line 33:
> > > > 31:
> > > > 32: typedef siginfo_t siginfo_t;
> > > > 33: typedef sigset_t sigset_t;
> > > 
> > > 
> > > I don't see why this is needed/wanted. We can include signal.h without a problem.
> > > I'm not even sure what these typedefs means ??
> > 
> > 
> > Coleen asked me to remove <signal.h> from the signals_posix.hpp, so those are forward declarations for the signal types we use.
> > I thought it was a reasonable request to minimize the number of headers. I saw some efforts in the past to cleanup header files, which is supposed to help with build times, so every little bit helps.
> 
> The only reason we would care is if signals_posix.hpp were included in
> many other handers/files and that should not be the case. This looks
> completely bogus to me as we need the types from signal.h in this header
> file. What do those typedefs even mean? I would expect a forward
> declaration to be of the form:
> 
> struct siginfo_t;
> 
> but you don't know what type sigset_t (could be integer or struct)
> actually is so you can't forward declare it that way.
> 

Oh you are right. Posix guarantees that siginfo_t and ucontext_t are structures; sigset_t can be either a structure or an integer type. Missed that.

I'd just either leave out the header. Or even leave it in. Do we have a policy like "all system headers only go into one central place, e.g. globalDefinitions"? Because if not, Gerard's original code would be more correct (where he included signal.h)

Just my 5 cent. I'll keep out of this discussion for now. Any form is fine for me.

Cheers Thomas

> David
> -----

dholmes-ora reviewed (approved):
Latest changes all look good.

Only open issue is the include, or not, of signal.h.

Thanks,
David

gerard-ziemski commented:
> _Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [hotspot-dev](mailto:hotspot-dev@openjdk.java.net):_
> 
> On 7/11/2020 2:40 am, Gerard Ziemski wrote:
> 
> > On Wed, 4 Nov 2020 04:22:05 GMT, David Holmes <dholmes at openjdk.org> wrote:
> > > > Gerard Ziemski has updated the pull request incrementally with two additional commits since the last revision:
> > > > - use ifdef(SIGDANGER) and ifdef(SIGTRAP)
> > > > - revert unblock_program_error_signals change
> > > 
> > > 
> > > src/hotspot/os/posix/signals_posix.hpp line 33:
> > > > 31:
> > > > 32: typedef siginfo_t siginfo_t;
> > > > 33: typedef sigset_t sigset_t;
> > > 
> > > 
> > > I don't see why this is needed/wanted. We can include signal.h without a problem.
> > > I'm not even sure what these typedefs means ??
> > 
> > 
> > Coleen asked me to remove <signal.h> from the signals_posix.hpp, so those are forward declarations for the signal types we use.
> > I thought it was a reasonable request to minimize the number of headers. I saw some efforts in the past to cleanup header files, which is supposed to help with build times, so every little bit helps.
> 
> The only reason we would care is if signals_posix.hpp were included in
> many other handers/files and that should not be the case. This looks
> completely bogus to me as we need the types from signal.h in this header
> file. 

Just trying to learn: why do **need** them? We only included them in the APIs here, but we don't actually **use** them otherwise.

And if we need to include `<signal.h>` then don't we also need to include the headers for `outputStream`, `Thread` and `OSThread`? All of these types are used to define the APIs in this header file and are used in the same capacity.


> What do those typedefs even mean? I would expect a forward
> declaration to be of the form:
> 
> struct siginfo_t;
> 
> but you don't know what type sigset_t (could be integer or struct)
> actually is so you can't forward declare it that way.

Forward declaration was a new concept to me, so I had to look it up and although it was easy enough to do it for the internal c++ classes, I struggled with the forward declaring of `struct sigset_t`, but I thought I found it.

Should I also drop the forward declaration for `outputStream`, `Thread`, `OSThread`?

tstuefe commented:
> > > Coleen asked me to remove <signal.h> from the signals_posix.hpp, so those are forward declarations for the signal types we use.
> > > I thought it was a reasonable request to minimize the number of headers. I saw some efforts in the past to cleanup header files, which is supposed to help with build times, so every little bit helps.
> > 
> > 
> > The only reason we would care is if signals_posix.hpp were included in
> > many other handers/files and that should not be the case. This looks
> > completely bogus to me as we need the types from signal.h in this header
> > file.
> 
> Just trying to learn: why do we **need** them? We only included them in the APIs here, but we don't actually **use** them otherwise.
> 
> And if we need to include `<signal.h>` then don't we also need to include the headers for `outputStream`, `Thread` and `OSThread`? All of these types are used to define the APIs in this header file and are used in the same capacity.

You need to include them if you:
- use the full type, so whoever compiles the header has to know the type size. E.g. if you pass a structure by value.
- use the type as pointer and do not forward declare it.

In hotspot, it is standard practice to forward-declare structures from some hotspot utility headers - eg ostream.hpp - to avoid including them. There is no guideline of when to do this, and obviously there is a point at which it is simpler and clearer to just include the header.

In my personal opinion forward declaration is just a bandaid to work around badly designed headers. Ideally, headers should be small and concise. But hotspot headers are balls of yarn. Pull one in you get a bunch of unrelated others. So people got used to forward declaring some classes instead, things like outputStream. I actually think its a bad practice and in the ideal world we would just include whatever we need.

Which also means that the benefit of forward-declaring types from system headers is limited. Posix headers are usually well designed, and you are better off just including them. Especially since you need to be careful here: it is not clear what these opaque posix types are actually. Sometimes the standard tells you: "The <signal.h> header shall define the siginfo_t type as a structure.." but sometimes it leaves it open: "sigset_t ... Integer or structure type ...".  

> 
> > What do those typedefs even mean? I would expect a forward
> > declaration to be of the form:
> > struct siginfo_t;
> > but you don't know what type sigset_t (could be integer or struct)
> > actually is so you can't forward declare it that way.
> 
> Forward declaration was a new concept to me, so I had to look it up and although it was easy enough to do it for the internal c++ classes, I struggled with the forward declaring of `struct sigset_t`, but I thought I found it.
> 
> Should I also drop the forward declaration for `outputStream`, `Thread`, `OSThread`?

For hotspot classes, I would leave the forward declarations in and the headers out. Current standard practice. System headers OTOH I would either include here or, somewhat better, in globalDefinitions_gcc.hpp.

gerard-ziemski commented:
> > > Forward declaration was a new concept to me, so I had to look it up and although it was easy enough to do it for the internal c++ classes, I struggled with the forward declaring of `struct sigset_t`, but I thought I found it.
> > > Should I also drop the forward declaration for `outputStream`, `Thread`, `OSThread`?
> > 
> > 
> > For hotspot classes, I would leave the forward declarations in and the headers out. Current standard practice. System headers OTOH I would either include here or, somewhat better, in globalDefinitions_gcc.hpp.
> 
> Ha! It is already in globalDefinitions_gcc.hpp so neither the direct
> include nor the forward declarations are actually needed.

Many thanks Thomas & David for the lesson on the header files!

If I understand it correctly, then we need to add `#include "utilities/globalDefinitions.hpp"` to signals_posix.hpp, correct?

tstuefe commented:
> > > > Forward declaration was a new concept to me, so I had to look it up and although it was easy enough to do it for the internal c++ classes, I struggled with the forward declaring of `struct sigset_t`, but I thought I found it.
> > > > Should I also drop the forward declaration for `outputStream`, `Thread`, `OSThread`?
> > > 
> > > 
> > > For hotspot classes, I would leave the forward declarations in and the headers out. Current standard practice. System headers OTOH I would either include here or, somewhat better, in globalDefinitions_gcc.hpp.
> > 
> > 
> > Ha! It is already in globalDefinitions_gcc.hpp so neither the direct
> > include nor the forward declarations are actually needed.
> 
> Many thanks Thomas & David for the lesson on the header files!
> 
> If I understand it correctly, then we need to add `#include "utilities/globalDefinitions.hpp"` to signals_posix.hpp, correct?

Yes.

gerard-ziemski commented:
> _Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [hotspot-dev](mailto:hotspot-dev@openjdk.java.net):_
> 
> On 11/11/2020 3:04 am, Gerard Ziemski wrote:
> 
> > Many thanks Thomas & David for the lesson on the header files!
> > If I understand it correctly, then we need to add `#include "utilities/globalDefinitions.hpp"` to signals_posix.hpp, correct?
> 
> It is already there.
> 
> #ifndef OS_POSIX_SIGNALS_POSIX_HPP
> #define OS_POSIX_SIGNALS_POSIX_HPP
> 
> #include "memory/allocation.hpp"
> #include "utilities/globalDefinitions.hpp"
> 
> #include <signal.h>
> 
> So you can just delete the include of signal.h

`#include "memory/allocation.hpp"` has:

`#include "memory/allStatic.hpp"
#include "utilities/globalDefinitions.hpp"`

so in the end `"memory/allocation.hpp"` is all we need.

tstuefe reviewed (changes requested):
Hi Gerard,

we are getting closer :)

Cheers, Thomas

dholmes-ora reviewed (approved):
I'm seeing unexpected changes relating to os::fetch_compiled_frame_from_context seemingly due to the "Merge from master".

Otherwise the actual described changes since last commit seem fine.

Once Thomas's last couple of comments are addressed this is done.

Thanks,
David

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.hpp:
```diff
@@ -28,49 +28,43 @@
 #include "utilities/globalDefinitions.hpp"
 
-#include <signal.h>
+// Forward declarations to be independent of the include structure.
```
Fixed.

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1374,31 +1386,11 @@ bool PosixSignals::is_sig_ignored(int sig) {
-#else
-   VMError::report_and_die("unimplemented ucontext_get_pc");
-#endif
+int PosixSignals::unblock_thread_signal_mask(const sigset_t *set) {
```
Fixed.

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1717,3 +1709,30 @@ void PosixSignals::do_resume(OSThread* osthread) {
+int PosixSignals::init() {
+  // initialize suspend/resume support - must do this before signal_sets_init()
+  if (SR_initialize() != 0) {
+    perror("SR_initialize failed");
```
Fixed, used `vm_exit_during_initialization(err_msg("SR_initialize failed"));`

tstuefe replied:
No need for the err_msg here, just use the plain string literal.

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1374,31 +1386,11 @@ bool PosixSignals::is_sig_ignored(int sig) {
-}
-
-void PosixSignals::signal_sets_init() {
+void signal_sets_init() {
```
Fixed.

gerard-ziemski commented:
> I'm seeing unexpected changes relating to os::fetch_compiled_frame_from_context seemingly due to the "Merge from master".
> 
> Otherwise the actual described changes since last commit seem fine.
> 
> Once Thomas's last couple of comments are addressed this is done.
> 
> Thanks,
> David

The `ucontext_get_pc()` API needs to be in `os::Posix`, not `PosixSignals`, so that change is OK.

gerard-ziemski commented:
Thank you Thomas and David for your reviews.

tstuefe commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1,4 +1,4 @@
-/*
+    /*
```
superfluous change

gerard-ziemski replied:
Fixed.

tstuefe commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1717,3 +1706,30 @@ void PosixSignals::do_resume(OSThread* osthread) {
+int PosixSignals::init() {
+  // initialize suspend/resume support - must do this before signal_sets_init()
+  if (SR_initialize() != 0) {
+    vm_exit_during_initialization(err_msg("SR_initialize failed"));
```
err_msg not needed, just use the plain literal

gerard-ziemski replied:
Ah, `err_msg()` is for the cases where we need to print out values (i.e. formatting).

tstuefe reviewed (commented):
Hi Gerard,

some final nits. 

I'll run the change through our system tonight to see if this broke one of our platforms. 

Cheers, Thomas

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1721,7 +1718,7 @@ void os::SuspendedThreadTask::internal_do_task() {
   // initialize suspend/resume support - must do this before signal_sets_init()
   if (SR_initialize() != 0) {
-    perror("SR_initialize failed");
+    vm_exit_during_initialization("SR_initialize failed");
```
You've lost the failure reason (errno) that perror would have provided. SR_initalize only returns non-zero if sigaction fails, which leaves errno set - hence perror() would report it. That said there are only two errors for sigaction (EFAULT or EINVAL) so we really haven't lost anything that useful.
Arguably the vm_exit... should be inside SR_initialize.
But I'm fine with this as-is so we can conclude this PR.

tstuefe replied:
Yes, but this relies on knowing the internal implementation of SR_initialize() (mainly that the last CRT call there was sigaction) and is different from all the other places where we do CRT calls. I'd rather add logging right there if the call fails like we do in other places.

tstuefe reviewed (approved):
AIX build went through, tests are fine so far. Looks good now. Ship it!

gerard-ziemski commented:
/integrate

gerard-ziemski commented:
Thank you David and Thomas for your reviews!
